dbms/src/Flash/Mpp/AsyncRequestHandler.h

```diff
@@ -0,0 +1,389 @@
+        }
+
+        GRPCReceiveQueueRes res = sendPackets();
+
+        if (!checkResultAfterSendingPackets(res))
+            return;
+
+        startAsyncRead();
+    }
+
+    void processWaitReWrite()
+    {
+        GRPCReceiveQueueRes res = reSendPackets();
+
+        if (!checkResultAfterSendingPackets(res))
+            return;
+
+        stage = AsyncRequestStage::WAIT_BATCH_READ;
+        startAsyncRead();
+    }
+
+    // return true if we still need to receive data from server.
+    bool checkResultAfterSendingPackets(GRPCReceiveQueueRes res)
+    {
+        if (unlikely(sendPacketHasError(res)))
+        {
+            closeConnection("Exchange receiver meet error : push packets fail");
+            return false;
+        }
+
+        if (unlikely(isChannelFull(res)))
+        {
+            asyncWaitForRewrite();
+            return false;
+        }
+
+        if (read_packet_index < batch_packet_count)
+        {
+            stage = AsyncRequestStage::WAIT_FINISH;
+            reader->finish(finish_status, thisAsUnaryCallback());
+            return false;
+        }
+
+        return true;
+    }
+
+    String getErrorFromPackets()
+    {
+        // step 1: check if there is error packet
+        // only the last packet may has error, see execute().
+        if (read_packet_index != 0 && packets[read_packet_index - 1]->hasError())
+            return packets[read_packet_index - 1]->error();
+        // step 2: check memory overflow error
+        for (Int32 i = 0; i < read_packet_index; ++i)
+        {
+            packets[i]->recomputeTrackedMem();
+            if (packets[i]->hasError())
+                return packets[i]->error();
+        }
+        return "";
+    }
+
+    bool retriable() const
+    {
+        return !has_data && retry_times + 1 < max_retry_times;
+    }
+
+    void closeConnection(String && msg)
+    {
+        if (!is_close_conn_called)
+        {
+            stage = AsyncRequestStage::FINISHED;
+            close_conn(!msg.empty(), msg, log);
+            is_close_conn_called = true;
+        }
+    }
+
+    void start()
+    {
+        stage = AsyncRequestStage::WAIT_MAKE_READER;
+
+        // Use lock to ensure async reader is unreachable from grpc thread before this function returns
+        std::lock_guard lock(make_reader_mu);
+        rpc_context->makeAsyncReader(request, reader, cq, thisAsUnaryCallback());
+    }
+
+    bool retryOrDone(String done_msg)
+    {
+        if (retriable())
+        {
+            ++retry_times;
+            stage = AsyncRequestStage::WAIT_RETRY;
+
+            // Let alarm put me into CompletionQueue after a while
+            // , so that we can try to connect again.
+            alarm.Set(cq, SystemClock::now() + std::chrono::seconds(retry_interval_time), this);
+            return true;
+        }
+        else
+        {
+            closeConnection(std::move(done_msg));
+            return false;
+        }
+    }
+
+    GRPCReceiveQueueRes sendPackets()
```

Comment:
```suggestion
    GRPCReceiveQueueRes sendPackets() noexcept
```

Reply:
> 

I think try-catch is better, and we can exist the tiflash when exception is thrown.

Reply:
ok